tests/test_cache.py

```diff
@@ -0,0 +1,246 @@
+import httpretty
+import json
+import os
+import pickle
+import redis
+import requests
+import shutil
+import struct
+import time
+import unittest
+from cachecontrol.caches.file_cache import FileCache
+from cachecontrol.caches.redis_cache import RedisCache
+from canonicalwebteam.http import CachedSession
+from datetime import timedelta
+from mockredis import mock_redis_client
+from unittest.mock import patch
+
+
+file_cache_directory = ".testcache"
+
+
+class MockRedisSingletons:
+    def __init__(self):
+        self.redis_clients = {}
+
+    def mock_redis_client_singleton(self, connection_pool, *args, **kwargs):
+
+        # A name has to passed into the connection pool, this will make sure
+        # that different Redis instances can be mocked
+        name = connection_pool.name
+
+        if name not in self.redis_clients:
+            self.redis_clients[name] = mock_redis_client()
+            self.redis_clients[name].name = name
+        return self.redis_clients[name]
+
+
+class TestCachedSession(unittest.TestCase):
+    mock_redis_singletons = MockRedisSingletons()
+
+    def tearDown(self):
+        try:
+            shutil.rmtree(file_cache_directory)
+        except:
+            pass
+
+    @httpretty.activate
+    def test_custom_heuristic(self):
+        def request_callback(request, uri, response_headers):
+            return [200, response_headers, json.dumps({"epoch": time.time()})]
+
+        httpretty.register_uri(
+            httpretty.GET, "https://now.httpbin.org", body=request_callback
+        )
+
+        session = CachedSession(
+            fallback_cache_duration=2,
+            file_cache_directory=file_cache_directory,
+        )
+
+        # with a 2s retention, and a 1.1s time between requests, 2 of the
+        # request should have the same epoch, where as the 3rd gets fresh data
+        # the first requests gets send at t=0
+
+        response_1 = session.get("https://now.httpbin.org")
+        time.sleep(1.5)
+        response_2 = session.get("https://now.httpbin.org")
+        time.sleep(0.7)
+        response_3 = session.get("https://now.httpbin.org")
+
+        self.assertEqual(response_1.text, response_2.text)
```

Comment:
using time.sleep can be tricky I would be careful with that:

```
======================================================================
FAIL: test_custom_heuristic (test_cache.TestCachedSession)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/tbmb/.pyenv/versions/3.6.5/envs/http/lib/python3.6/site-packages/httpretty-0.9.6-py3.6.egg/httpretty/core.py", line 1702, in wrapper
    return test(*args, **kw)
  File "/home/tbmb/Documents/projects/http/tests/test_cache.py", line 71, in test_custom_heuristic
    self.assertEqual(response_1.text, response_2.text)
AssertionError: '{"epoch": 1548239344.9905274}' != '{"epoch": 1548239346.5064354}'
- {"epoch": 1548239344.9905274}
?                     ^^^^ --
+ {"epoch": 1548239346.5064354}
?                    +++++ ^


----------------------------------------------------------------------
```

Reply:
As discussed I have changed this to use [freezegun](https://github.com/spulec/freezegun). It is much more reliable and quicker :)